src/core/Common/FunctionalUnit.ts

```diff
@@ -14,104 +14,225 @@ export let FunctionalUnitTypeNames: string[] = ['Integer Sum', 'Integer Multiply
 
 export const FUNCTIONALUNITTYPESQUANTITY = FunctionalUnitType.JUMP - FunctionalUnitType.INTEGERSUM + 1;
 
-export class FunctionalUnit {
+export interface FunctionalUntitVisualEntry {
+    id: number;
+    value: string;
+    uuid: number;
+}
 
-    private _status: Status;
+const FunctionalUnitLantencies: Record<FunctionalUnitType, number> = {
+    [FunctionalUnitType.INTEGERSUM]: 1,
+    [FunctionalUnitType.INTEGERMULTIPLY]: 2,
+    [FunctionalUnitType.FLOATINGSUM]: 4,
+    [FunctionalUnitType.FLOATINGMULTIPLY]: 6,
+    [FunctionalUnitType.MEMORY]: 4,
+    [FunctionalUnitType.JUMP]: 2
+};
+
+export const FunctionalUnitNumbers: Record<FunctionalUnitType, number> = {
+    [FunctionalUnitType.INTEGERSUM]: 2,
+    [FunctionalUnitType.INTEGERMULTIPLY]: 2,
+    [FunctionalUnitType.FLOATINGSUM]: 2,
+    [FunctionalUnitType.FLOATINGMULTIPLY]: 2,
+    [FunctionalUnitType.MEMORY]: 2,
+    [FunctionalUnitType.JUMP]: 1
+};
+
+export interface FunctionalUnitResult {
+    instruction: Instruction;
+    result: number;
+    ref: number; // TODO: use instruction uid
+}
+
+interface FunctionalUnitInstruction {
+    instruction: Instruction;
+    ref: number; // TODO: use instruction uid
+    blankTimeUnitsAhead: number; // how many time units ahead are blank, i.e. how many time units need to pass before the instruction can be executed
+}
+
+export class FunctionalUnit {
 
     private _type: FunctionalUnitType;
     private _latency: number;
-    private _flow: Instruction[];
-
-    constructor() {
-        this._flow = null;
-        this._status = new Status();
-        this._status.lastInstruction = 0;
-        this._status.stall = 0;
-        this._status.instructionNumber = 0;
-    }
+    private _stalled: number; // if >0, it is stalling (for ex because a memory fail) for that many cycles
+    private _instructions: FunctionalUnitInstruction[];
 
-    public get status(): Status {
-        return this._status;
-    }
-
-    public set status(value: Status) {
-        this._status = value;
-    }
+    private _nextRef: number; //TODO: use instruction uid
+    private _currentBlankTimeUnits: number;
+    private _hasExectutedInstBeforeTick: boolean;
 
     public get type(): FunctionalUnitType {
         return this._type;
     }
 
-    public set type(value: FunctionalUnitType) {
-        this._type = value;
-    }
-
     public get latency(): number {
         return this._latency;
     }
 
-    public set latency(value: number) {
-        this._latency = value;
-        this._status.lastInstruction = value - 1;
-        this._status.instructionNumber = 0;
-        this._flow = new Array(value);
-        this._flow.fill(null);
-    }
+    constructor(type: FunctionalUnitType, latency: number = FunctionalUnitLantencies[type]) {
+        this._type = type;
+        this._latency = latency;
+        this.clean();
+    }
+
+    public clean() {
```

Comment:
> The existence of clean in all the units is because there should be a clear difference between rebuilding the entire unit or just wanting to clean or reset it.

The result should be the same, but reinstanciating is explicit on what the object will turn into, while `clean()` isn't. If you are looking for a semantic difference, I can understand that, then you can do:

```
clean = (): FunctionalUnit => {
  const cleanedInstance = new FunctionalUnit(this._type, this._latency);
  return cleanedInstance;
}
```

But, in this case, the `clean` terminology is confusing in general and, since something like `nodiscard`(1) does not exist in TypeScript, this method can be misused easily.

(1): https://en.cppreference.com/w/cpp/language/attributes/nodiscard


> This is very important, as changing or resseting a unit construction parameter during execution or without a "hard" reset may cause some strange side effects.

I'm not sure this is a thing. Can you elaborate?

> So, re-instantiation might also prone to errors, an unaware developer might clean a unit without copying it old constructions parameters or vice versa.

An unaware developer might not instantiate anything at all anywhere in the code. This is why testing is done. And, in this case, there is more testing load for the cases where `clean()` is implemented and used, because it implicitly modifies state. Also, what do you mean with _vice versa_ in this context?

> But last and most important, a clean function looks more clean than re-instantiating.

Looking clean does not have to be a positive thing depending on what you are sacrificing for that cleanliness. And, again, in this case, that is explicitness.

> While we currently, at most, we have as much as 2 o 3 construction parameters, in not so distant future new machines with new units will be added or old units will be made more flexible. For example, if the functional unit gets parametrized to 5 parameters, which looks more clean?

This is what, in our sector, we call _"solving imaginary issues"_. It's a problem we don't have. Now, let's suppose we have it. Still, we can be doing this:

```
class FunctionalUnit {
  ...

  public static const fromFunctionalUnit(
    functionalUnit: FunctionalUnit
  ): FunctionalUnit {
  
  const newFunctionalUnit = new FunctionalUnit(functionalUnit.type, functionalUnit.latency);
  return newFunctionalUnit;
  }
}
```

You might think: _"Well, but this is less implicit that just using the constructor! And it would need a test, as well. How does the developer know which parameters is this using for instantiation?"_

You would be mostly right by thinking that. It is definitely less implicit, and would need a test. We could then discuss on its naming, if it can get confused as a _deep copy_... At that point, the best solution is for this method to not exist, and for the developer to bootstrap the functionality as a helper in their own code, making it explicit for them.

But, after all, this issue is still imaginary, and the focus should be in something else.

Reply:
Firstly, yes, I fully agree that using the name clean is, at least, confusing. And this behavior can be easily ignored by others dev.

Before going point by point, I understand the point that `clean` modifies the state implicit, but that is the whole of these classes. Like the same things we are arguing for clean can be argued for `tic`:
- It modifies the internal state of the unit, causing side effects
- An unaware dev can ignore or, better yet, call it in the wrong machine execution step and cause all sort of strange behaviors and bugs
This can be said for a lot of methods of these classes, because in part, these reduce the verbosity of _Superescalar.ts_ in favor of readability.


> I'm not sure this is a thing. Can you elaborate?
For example, if the register size is changed, but the machine is not "hard" reset, old code with now can go out of registers bound can be executed as is already loaded. Another example would be during an execution with custom Functional Units latencies, if the miss predicts code get modified and when cleaning these units by re-instantiating the old latency does not get copied, from that point, the entire execution will be wrong. The problem is that unless the user was very vigilant, they will have 0 clues that the machine is giving a wrong result.

> An unaware developer might not instantiate anything at all anywhere in the code. This is why testing is done. And, in this case, there is more testing load for the cases where clean() is implemented and used, because it implicitly modifies state. 
Can you elaborate on this?

> Also, what do you mean with vice versa in this context?
Sorry for my laziness and just putting "vice versa", I was referring to that an unaware developer might copy old construction parameters of a unit when the machine is resetting to it default parameters, so those parameters should have returned to their default. 

> Looking clean does not have to be a positive thing depending on what you are sacrificing for that cleanliness. And, again, in this case, that is explicitness.
This is an interesting point I would like to talk, but I think it's better to do on a meeting.

> This is what, in our sector, we call "solving imaginary issues". 
In the current status of this refactor, totally fair, but I want to clarify that this has to do with a related topic that is more or less important and I also want to discuss with you.

> Still, we can be doing this:
> 
> class FunctionalUnit {
>   ...
> 
>   public static const fromFunctionalUnit(
>     functionalUnit: FunctionalUnit
>   ): FunctionalUnit {
>   
>   const newFunctionalUnit = new FunctionalUnit(functionalUnit.type, functionalUnit.latency);
>   return newFunctionalUnit;
>   }
> }
> You might think: "Well, but this is less implicit that just using the constructor! And it would need a test, as well. How does the developer know which parameters is this using for instantiation?"
> 
> You would be mostly right by thinking that. It is definitely less implicit, and would need a test. We could then discuss on its naming, if it can get confused as a deep copy... At that point, the best solution is for this method to not exist, and for the developer to bootstrap the functionality as a helper in their own code, making it explicit for them.
> 
> But, after all, this issue is still imaginary, and the focus should be in something else.

> If you are looking for a semantic difference, I can understand that, then you can do:
> 
> clean = (): FunctionalUnit => {
>   const cleanedInstance = new FunctionalUnit(this._type, this._latency);
>   return cleanedInstance;
> }

Those forms are just complicating things up.

Reply:
Btw, sorry if I'm being too stubborn. I have to admit now I'm also not very convinced of the clean method. It has another problem that I ignored: A dev can add a new private variable to a unit and forgot to clean in the method.

I will try to implement your method, but I'm still convicted that is too verbose and have the similar problems than clean.